yesqa.py

```diff
@@ -118,14 +118,11 @@ def fix_file(filename: str) -> int:
     if src_no_comments == contents_text:
         return 0
 
-    with tempfile.NamedTemporaryFile(
-        dir=os.path.dirname(filename),
```

Comment:
hmmmm shoot sorry, now that I read this a little more closely it looks like this intended to write the file to the same directory that it already existed in (presumably so a flake8 plugin can figure out import paths if needed)

so we probably need to do this instead:

```python
filename = tempfile.mkstemp(dir=..., prefix=..., suffix=...)
try:
    with open(filename, 'wb') as f:
        f.write(src_no_comments.encode())
    flake8_results = _run_flake8(tmpfile.name)
finally:
    os.remove(tmpfile)
```

Reply:
err that's not right

maybe this

```python
    try:
        with tempfile.NamedTemporaryFile(
            dir=os.path.dirname(filename),
            prefix=os.path.basename(filename),
            suffix='.py',
            delete=False,
        ) as tmpfile:
            tmpfile.write(src_no_comments.encode())
        flake8_results = _run_flake8(tmpfile.name)
    finally:
        os.remove(tmpfile.name)
```

Reply:
This should more or less the same as https://github.com/asottile/yesqa/issues/45#issuecomment-684096528 (besides the flake8_results being outside of the context manager).
Since `TemporaryFile` is pretty much a convenience layer around `mkstemp` or?

Reply:
I guess, mkstemp is kinda clunky

```python
fd, path = tempfile.mkstemp(...)
try:
    os.fdclose(fd)
    with open(path, 'wb') as f:
        f.write(src_no_comments.encode())
    flake8_results = _run_flake8(tmpfile.name)
finally:
    os.remove(path)
```

Reply:
actually, that's probably how it has to happen (with `mkstemp`) otherwise there's a potential case where `NamedTemporaryFile` raises and it tries to clean up something which doesn't exist

Reply:
Just tested it at the project I was setting up `pre-commit` for and it did work fine (no false positive removals).
But the project doesn't use flake8-plugin + the `noqa`'s all concern `E402` and `F401`, so this might not be a good implementation test.

Reply:
So I did some digging and it looks like that the problem with `NamedTemporaryFile` on windows is due to the way cleanup is handeled on windows.
On windows a special flag is set, to tell windows the file should be deleted on close:
https://github.com/python/cpython/blob/e55a0e971b4bf9e6473bf0ca6bebdff76c075ef6/Lib/tempfile.py#L536-L539

On Posix systems, `os.unlink` is called in case `delete==True`:
https://github.com/python/cpython/blob/e55a0e971b4bf9e6473bf0ca6bebdff76c075ef6/Lib/tempfile.py#L433

The cleanup in case of an error is done with `except BaseException:`, which should be equvalent to `try: finally:` :
https://github.com/python/cpython/blob/e55a0e971b4bf9e6473bf0ca6bebdff76c075ef6/Lib/tempfile.py#L547-L550

Looking at the source code of `NamedTemporaryFile`, the `os.fdclose` might not be needed.

Reply:
look closely :) I only used fdclose for mkstemp

Reply:
What I ment was that `mkstemp` returns `_mkstemp_inner`:
https://github.com/python/cpython/blob/e55a0e971b4bf9e6473bf0ca6bebdff76c075ef6/Lib/tempfile.py#L332

And `NamedTemporaryFile` uses `_mkstemp_inner` internally, without the `os.fdclose`:
https://github.com/python/cpython/blob/e55a0e971b4bf9e6473bf0ca6bebdff76c075ef6/Lib/tempfile.py#L541-L544

Anyway, IMHO `NamedTemporaryFile` + `try: finally:`, looks prettier anyway and it resembles the original code closer.
But since aesthetics are always at least somewhat personally, the decision is up to you ^^

Reply:
the problem with `NamedTemporaryFile` + `try: finally:` is there will require an if statement, and that if statement will be impossible to cover

Reply:
Ahhh now I also get why the `mkstemp` is outside of the `try: finally:` 😅

Reply:
Oh and `os.fdclose` isn't neeeded if the file is opened with `open(fd, 'wb')` instead of `open(path, 'wb')` 😄